### PR TITLE
Rework nodePort gateway service

### PR DESCRIPTION
**Pull request by hzxuzhonghu**

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
cc @rshriram

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 1]
```diff
@@ -174,31 +174,31 @@ func (c *Controller) GetService(hostname host.Name) (*model.Service, error) {
 			// VIPs or CIDR ranges in the address field
 			return service, nil
 		}
+
 		// This is K8S typically
 		if out == nil {
-			out = service
+			out = service.DeepCopy()
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/aggregate/controller.go:180]: Always a deep copy?

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 2]
```diff
@@ -407,10 +402,9 @@ func (c *Controller) onNodeEvent(obj interface{}, event model.Event) error {
 	return nil
 }
 
-func isNodePortGatewayService(svc *model.Service) bool {
-	svc.Mutex.RLock()
-	defer svc.Mutex.RUnlock()
-	return svc.Attributes.ClusterExternalPorts != nil
+func isNodePortGatewayService(svc *v1.Service) bool {
+	_, ok := svc.Annotations[kube.NodeSelectorAnnotation]
+	return ok && svc.Spec.Type == v1.ServiceTypeNodePort
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:407]: Should we nuke this entire thing and simply collect node ips from the pods of the gateway?

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 3]
```diff
@@ -529,36 +523,49 @@ func (c *Controller) GetService(hostname host.Name) (*model.Service, error) {
...
+			var extAddresses []string
+			for _, n := range c.nodeInfoMap {
+				extAddresses = append(extAddresses, n.address)
+			}
+			svc.Attributes.ClusterExternalAddresses = map[string][]string{c.clusterID: extAddresses}
+		} else {
+			var nodeAddresses []string
+			for _, n := range c.nodeInfoMap {
+				if nodeSelector.SubsetOf(n.labels) {
+					nodeAddresses = append(nodeAddresses, n.address)
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:563]: If we collect the node ips from the pods, all this logic goes away.

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:407]: I have thought about when i wrote this pr and found that it is a little complicated.  I will work as a separate one.  This is more like an optimization

File: pilot/pkg/serviceregistry/aggregate/controller.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/aggregate/controller.go:180]: Yes, it doesn't matter much as currently is it only used for gateway service.

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:563]: yes

**Pull request merged**